### ISSUE TITLE
1837 Bugs and Improvements

### ISSUE THREAD

**Issue by neutronc** [Rails-18xx/Rails#329]
[Rails-1.zip](https://github.com/Rails-18xx/Rails/files/5813319/Rails-1.zip)
Edit: numbered the issues (EV)

 General impression is that the game is playable somewhat
farther than I got the last time I worked on the code.
Someone else has worked on it a moderate amount, but has by
no means finished it.

[1]  Minor bug: The Game Status screen shows most of the
companies but omits two of the three state companies, the Sd
and Ug.  All three should be shown.
**Update**: fixed. Was a config typo.

[2]  The use of multiple "start rounds" is a choice I made when
I worked on it, but I now believe it was a mistake.
Technically the only the first screen of that type is the
"start packet," defined in the rules as "it all has to sell
out before an operating round can take place."

The way I should have handled the sale of the K1-K3 and U1-U3
is to make those the first companies that are available, using
the existing mechanism of "layers" that limits the order of
starting new companies in 1829, 1835, and 1853.
**Update** EV: willl do.

[3]  Feature request: When using the three-column start packet,
show it graphically alongside the start round screen.  As it is,
there is no way to tell which shares are above which others (and
thus will make them accessible when purchased).
**Update** EV: low priority.

[4]   Bug (as I see it): The window label does not say "start round
(number)" even if more than one "start round" takes place.
(Which can happen in other games, for instance 1830 and 1835, if
an operating round occurs before all start packet contents have
been purchased.)
**Update**: done.

[5]  Bug: In my test game today, all of the K1-K3 sold out before
any operating round took place, but of the U1-U3, only the U1
president's share sold.  We played one operating round.  When it
ended the "start round" window displayed again, to sell the rest
of U1-U3; but only one player had enough money to buy any of the
shares.  That player bought one, and the other two players passed
-- and then Rails ended the round before the player with money
could buy another share!
   Save file:  1837_20210113_2138_OperatingRound 0.1.rails
**Update**: probably related to [2].

[6]  Worse bug: The "forced undo" mechanism would NOT take me back
into that "start round" to let the player buy more shares.  When
I tried it, the "Leave history" and "Play from here" buttons
spontaneously disappeared from the Game Report window.
**Update**: probably related to [2].

[7]  Feature request: Show a confirmation dialog to the owner of a
coal company if he tries to end the company's turn in OR1 without
buying a train.  While not required, this is a nasty mistake you
usually can't recover from.
**Update** EV: doesn't this apply to all 18xx minors where train ownership is optional? Low priority.

[8]  Cosmetic bug: Coal mine hexes ought to look very different.
(1) Move the company name out of the center of the hex where the
track makes it unreadable.  Move it to the opposite side of the
center of the hex from the track.  (2) Make the "mine station
marker" square, not round, so it doesn't appear to be a city.
(3) Show the coal mine revenue as if it were a red offboard hex,
including the higher revenue that begins in phase 5.  All these
things should be done before the companies are even purchased in
the start auction because the person considering buying them
needs the info.
**Update** EV: can be fixed by a background map. Anyone?

[9]   For the same reason, the (i) button beside each company on the
start round screen should identify the major company into which
each coal company must eventually merge.  This is important,
especially for EKT which must merge into MS but usually does not
build its track near MS.
**Update** Done. The company info from the OR window Info menu is exactly copied to the corresponding i-button.

[10] Bug: Coal mine income is being paid out as part of the
company's train run.  It is supposed to be direct revenue, like
the mail contracts in 1817.  This is also true when a major
company obtains a G train and runs it to a mine.
**Update** EV: Done. The solution may be a kludge, but it seems to work (not yet  tested for non-coal companies).
What you would really want is the capability to have a "Revenue" object containing two (or more) revenue components (integers), and pass that around in stead of a single revenue integer.

[11]  Save file:  1837_20210113_2153_OperatingRound 0.1.rails
(1) Bug: In this test game, K1 is collecting 100 for its run with
one 2-train!  Its run ought to end at the dit in F18.  
**Update**: fixed, was config error.

(2) Passenger trains must count the dits, 
**Update**: fixed, was config error.

(3) and may not run to mines.
**Update** EV: fixed. It required a fundamental upgrade, needed in several other games, like 1844 and 18VA.

[12]   Bug: Building on a private company's hex now works in two ways
that disobey the 1837 rules (in my opinion).  
(1) Unlike some other games where privates have "teleport" powers, using a
private company to build track in 1837 does not exempt you from
the normal requirement to be connected to that track.  
**Update**: fixed, was config error.

(2) During phase 2 it is NOT legal for a non-owner of a private company to
build track on that private company's hex AT ALL, even by paying
the terrain cost.  You have to wait until phase 3.
**Update**: fixed.

[13]   Subtle bug: Paul Work's translation of the rules (which I can't
find right now) said that unlike most other games, a company is
allowed to backtrack when tracing a route from an existing station
of the company in order to place another station (but only once
per station being placed).
**Update**: a solution is not in sight. Best chance may be to remove the token laying connectedness requirement entirely.

[14]  Cosmetic bug: K1's home station is mislabeled as KK.  That
station is K1 until it merges into the KK.
**Update**: see [8].

The following issues have been added later by Erik:

[15] After creating a route J34 (ZKB mine) - J36 (SB) - I35 - J34 (town) - I33 - I31 - J30 - K29 (SPB mine), two errors show up:
(1) a G1 train takes the whole route, including both mines, which is not allowed
**Update**: fixed.

(2) the mine and the town on J34 both add 25G to the revenue. The town should add 10G.
Both problems mean that a more fine-grained approach to the values and access properties is needed, so I'll have to revisit the Stop properties concept. See also the attachment to issue #234; it's about time to create a wiki page for that subject.
**Update** EV: Fixed. A phase-dependent hex value does no longer override a nonzero fixed station value as defined on the tile. 

[16] Coal companies may only buy goods trains. This has apparently not been implemented yet.
**Update**: done.

[17] Goods trains must include one mine. This has apparently not been implemented yet.
**Update**: done.

[18] Wien (Vienna) cannot be upgraded to green. The connection is seen, but the upgrade is disallowed (pink border).
**Update**: This was a hard one. It turned out, that `TileUpgrade.assignStations()` could not deal with the six-fold symmetry of tile 427: all rotations of this tile are equivalent and thus valid, but that method found none, and thus no stationmapping!  As I don't understand what that method is doing, I had to add my own code to fix it in the obvious way: for each side, map the old and new stations having tracks to that side. I have also implemented the option to prescribe a fixed tile orientation, and used it for tiles 427 and 436 (the latter would need it anyway).

[19] (Removed, was not a bug).

[20] The 3+1-train is not released when the 3-trains are sold out.
**Update**: Fixed. Problem was caused by the 2G-trains being configured in between these two types.
All G-trains have been moved to the end of the list. Refactored the train type release code to include the option that the initial train type (2) can also release other types (1G in this case). 
Note: from Lonny's 2015 version (2.0) of the English rules, chapter XI,  I conclude, that it is not so, that selling out the 1G trains releases the 2G ones, etc. Confirmed that Rails does not do so.
 
[21] Trade-ins. Any company that already has the maximum number of trains it may own, but wants to buy a new train, may pay the bank to "scrap" one of its existing trains and buy the new train. (The new train purchase is a condition of scrapping.) The fee for scrapping a train is half its original new purchase price. The scrapped train is placed in the bank pool and may be purchased again. A company may not scrap a train in order to buy (1) any previously scrapped train or (2) another train of the same type as the one being scrapped. I am not clear about whether a company making a forced train purchases may have to buy the scrapped train if a more expensive new train is available.
**Update**: Done. A new stub was added to allow subclasses to fill in the details of train exchange when at the train limit. The existing 1830-style and 18EU (Pullmann) exchanges have _not_ (yet) been integrated with this new provision (if that would be possible at all).

**Comment by erik-vos** [comment]
That's a lot of work to do! A few comments:

- Multiple start rounds: agreed.
- Start round number: is already maintained by GameManager, only needs to be picked up by the UI. See StatusWindow.updateStatus() for how it's done for stock rounds.
- Confirmation dialog: no big deal, should IMO be done in ORUIManager or its subclass.
- Coal company info ("i" buttons): the Info menu in the OR screen already has extensive info. Don't know why only part of that is copied to the StartRound window (see e.g. AB), but adding more stuff shouldn't be a big deal.
- Counting towns: remove "countTowns="no" from Defaults in Game.xml, or specify "yes" per applicable train type.
- No teleports: add "connected="yes"" to the special power config.
- Not run to mines: I think the generic feature to configure train type specific exclusions still has to be developed. Is on my to-do list, because 18VA does need it too.

**Comment by erik-vos** [comment]
While working at 1837, more bugs surface. I have added [15] to the top comment of this issue.

**Comment by erik-vos** [comment]
To the top comment I have added my view on the status of all points raised in this issue, see there.

**Comment by erik-vos** [comment]
Making good progress.

The basic requirement (also for 1844) that only certain trains may run to a certain station (e.g. a mine) has been addressed by adding a 4th RunTo option, to be spelled as "trains:goods". Here "goods" is a train **category**, to be added to the definition of any train type that is not a passenger train (which remains the default). Values other than "passenger" are not predefined and free to choose. 

If anyone knows a more railwayish name for the **category** concept, please let me know.

I am working on a new Wiki page that explains the configuration of this and other Stop properties. The introductory sections are about finished and can be viewed (part 3 of the Configuration details series).

**Comment by erik-vos** [comment]
See the opening comment of this issue for updates.
As I think that 1837 is now a lot better playable than a week ago, I'm considering to create a PR so others can have a look. Does that make sense?

**Comment by neutronc** [comment]
Sure we can release a beta version of 2.4 with the fixes anytime

**Comment by jdgalt** [comment]
Re. [22] - This is a needed feature arising from the two types of trains, and also arises in 1853.
The correct behavior, as I understand it, is as follows.
*  1G and 2 trains are available right away.  (But just listing the 1G's first should work, because unless someone makes the big mistake discussed in [7], all ten 1G trains will have been sold before the first run of any company that is allowed to have passenger trains.)
*  Passenger trains of each later type become available when the preceding passenger train type sells out, regardless of any goods trains.
*  2G trains become available after the first 3 is purchased.  I am not sure whether all 1Gs must also be purchased before the first 2G, but again that issue is unlikely to come up.
*  3G trains become available after the first 4 is purchased.  I am not sure whether all 2Gs must also be purchased before the first 3G.
*  4G trains become available after the first 5 is purchased.  I am not sure whether all 3Gs must also be purchased before the first 4G.

While we're at it, I'll point out one more feature that's in the 1837 rules but not in the code yet:

*  Trade-ins.  Any company that already has the maximum number of trains it may own, but wants to buy a new train, may pay the bank to "scrap" one of its existing trains and buy the new train.  (The new train purchase is a condition of scrapping.)  The fee for scrapping a train is half its original new purchase price.  The scrapped train is placed in the bank pool and may be purchased again.  A company may not scrap a train in order to buy (1) any previously scrapped train or (2) another train of the same type as the one being scrapped.  I am not clear about whether a company making a forced train purchases may have to buy the scrapped train if a more expensive new train is available.

**Comment by erik-vos** [comment]
On the G trains: what you describe is how it now works, after moving all G trains to the end of the list. Or at least how it _should_ work, because my testing has not yet progressed past the 3-trains.
Sure I could have left the 1G at the top, but my way I could drop the artificial phase change after buying the first 1G.

I'll add the trade-ins to the list. No doubt there is more.

**Comment by erik-vos** [comment]
Update on the G-trains: as it is in Rails now, selling out the 1Gs does _not_ release the 2G-trains. So I expect the other G-trains behave similarly. 
I still think that is correct. Just found the draft v3.0 rules for AAG of 2019, and these are also silent on this matter. The only German version I found is an old single blue sheet that outlines the differences with 1835 - same thing.

**Comment by erik-vos** [comment]
Have done (sort of) separating the mine revenue and pay all of it out to the company.
The solution may be a kludge, but it seems to work (not yet tested for non-coal companies).
What you would really want is a "Revenue" object containing two (or more) revenue components (integers), and pass that around, in stead of a single revenue integer.
Will now continue with issues [12.2] and [21].

If anyone wants to have a look, I can push the current code to branch 1837 of my fork.
